Detectors/MUON/MCH/Tracking/src/TrackExtrap.cxx

```diff
@@ -329,6 +546,256 @@ void TrackExtrap::addMCSEffect(TrackParam* trackParam, double dZ, double x0)
+
+  // Get the covariance matrix in the (XVtx, X, YVtx, Y, q*PTot) coordinate system
+  TMatrixD paramCovVtx(5, 5);
+  paramCovVtx.Zero();
+  paramCovVtx(0, 0) = errXVtx * errXVtx;
+  paramCovVtx(1, 1) = paramCovP(0, 0);
+  paramCovVtx(2, 2) = errYVtx * errYVtx;
+  paramCovVtx(3, 3) = paramCovP(2, 2);
+  paramCovVtx(4, 4) = paramCovP(4, 4);
+  paramCovVtx(1, 3) = paramCovP(0, 2);
+  paramCovVtx(3, 1) = paramCovP(2, 0);
+  paramCovVtx(1, 4) = paramCovP(0, 4);
+  paramCovVtx(4, 1) = paramCovP(4, 0);
+  paramCovVtx(3, 4) = paramCovP(2, 4);
+  paramCovVtx(4, 3) = paramCovP(4, 2);
+
+  // Jacobian of the transformation (XVtx, X, YVtx, Y, q*PTot) -> (XVtx, SlopeXVtx, YVtx, SlopeYVtx, q*PTotVtx)
+  TMatrixD jacob(5, 5);
+  jacob.UnitMatrix();
+  jacob(1, 0) = -1. / (zB - zVtx);
+  jacob(1, 1) = 1. / (zB - zVtx);
+  jacob(3, 2) = -1. / (zB - zVtx);
+  jacob(3, 3) = 1. / (zB - zVtx);
+
+  // Compute covariances at vertex in the (XVtx, SlopeXVtx, YVtx, SlopeYVtx, q*PTotVtx) coordinate system
+  TMatrixD tmp(paramCovVtx, TMatrixD::kMultTranspose, jacob);
+  TMatrixD newParamCov(jacob, TMatrixD::kMult, tmp);
+
+  // Compute covariances at vertex in the (XVtx, SlopeXVtx, YVtx, SlopeYVtx, q/PyzVtx) coordinate system
+  covP2Cov(newParam, newParamCov);
+
+  // Set parameters and covariances at vertex
+  param->setParameters(newParam);
+  param->setZ(zVtx);
+  param->setCovariances(newParamCov);
+
+  return true;
+}
+
+//__________________________________________________________________________
+void TrackExtrap::correctELossEffectInAbsorber(TrackParam* param, double eLoss, double sigmaELoss2)
+{
+  /// Correct parameters for energy loss and add energy loss fluctuation effect to covariances
+
+  // Get parameter covariances in (X, SlopeX, Y, SlopeY, q*PTot) coordinate system
+  TMatrixD newParamCov(param->getCovariances());
+  cov2CovP(param->getParameters(), newParamCov);
+
+  // Compute new parameters corrected for energy loss
+  double p = param->p();
+  double e = TMath::Sqrt(p * p + SMuMass * SMuMass);
+  double eCorr = e + eLoss;
+  double pCorr = TMath::Sqrt(eCorr * eCorr - SMuMass * SMuMass);
+  double nonBendingSlope = param->getNonBendingSlope();
+  double bendingSlope = param->getBendingSlope();
+  param->setInverseBendingMomentum(param->getCharge() / pCorr *
+                                   TMath::Sqrt(1.0 + nonBendingSlope * nonBendingSlope + bendingSlope * bendingSlope) /
+                                   TMath::Sqrt(1.0 + bendingSlope * bendingSlope));
+
+  // Add effects of energy loss fluctuation to covariances
+  newParamCov(4, 4) += eCorr * eCorr / pCorr / pCorr * sigmaELoss2;
+
+  // Get new parameter covariances in (X, SlopeX, Y, SlopeY, q/Pyz) coordinate system
+  covP2Cov(param->getParameters(), newParamCov);
+
+  // Set new parameter covariances
+  param->setCovariances(newParamCov);
+}
+
+//__________________________________________________________________________
+void TrackExtrap::cov2CovP(const TMatrixD& param, TMatrixD& cov)
```

Comment:
this comment is probably true for most of the functions in this class, but also probably easier to implement for this one : what about a (simple) unit test to check the correctness of that method ?

Reply:
will think about it...